src/other/broadcasting.jl

```diff
@@ -1,3 +1,63 @@
+### Broadcasting
+
+Base.getindex(df::AbstractDataFrame, idx::CartesianIndex{2}) = df[idx[1], idx[2]]
+Base.setindex!(df::AbstractDataFrame, val, idx::CartesianIndex{2}) =
+    (df[idx[1], idx[2]] = val)
+
+Base.broadcastable(df::AbstractDataFrame) = df
+
+struct DataFrameStyle <: Base.Broadcast.BroadcastStyle end
+
+Base.Broadcast.BroadcastStyle(::Type{<:AbstractDataFrame}) =
+    DataFrameStyle()
+
+Base.Broadcast.BroadcastStyle(::DataFrameStyle, ::Base.Broadcast.BroadcastStyle) = DataFrameStyle()
+Base.Broadcast.BroadcastStyle(::Base.Broadcast.BroadcastStyle, ::DataFrameStyle) = DataFrameStyle()
+Base.Broadcast.BroadcastStyle(::DataFrameStyle, ::DataFrameStyle) = DataFrameStyle()
+
+function copyto_widen!(res::AbstractVector{T},
+                       bc::Base.Broadcast.Broadcasted{DataFrameStyle},
+                       pos, col) where T
+    for i in pos:length(axes(bc)[1])
+        val = bc[CartesianIndex(i, col)]
```

Comment:
I do not think it is possible easily, unless @mbauman has an idea how to do it safely (essentially we would have to break down a 2-dimenstional `bc` into a collection of 1-dimensional columnwise `bc`s that would select appropriate columns from data frames taking part in the operation).

Reply:
Right. Maybe we can special-case data frames? Anyway that's the only type for which it matters.

Reply:
What I imagine you would have to do is the following:
1. flatten `bc`
2. manually check dimensionality of all `args` and which of them is `AbstractDataFrame` (I already do this for)
3. pass to flattened function arguments by "manual" broadcast of all `args`

This is doable but complex (effectively you have to rewrite broadcasting machinery), unless @mbauman could propose something better.

If what I write above is the best solution I would:
1. leave the implementation as is
2. open an issue/PR for performance improvements as to make the above fully efficient is not an easy task (then we can safely merge the faster approach when we are sure it is fully safe).

Reply:
Well, you could just re-use broadcast here.  Just write a `preprocess`-like recursive function that hoists the columns out.  In other words (pseudo-pseudo code, untested, etc):

```julia
import Broadcast: Broadcasted, extrude, broadcast_unalias
@inline preprocess(dest, bc::Broadcasted{Style}, col) where {Style} = Broadcasted{Style}(bc.f, map(x->preprocess(dest, x, col), bc.args), bc.axes)
preprocess(dest, x, col) = extrude(broadcast_unalias(dest, x))
preprocess(dest, x::AbstractDataFrame, col) = extrude(broadcast_unalias(dest, x[:,col]))
```

Then:
```julia
# ...
    for j in axes(bc)[2]
        bccol = preprocess(dest, bc, j)
        inner(dest, bccol, j)
    end
# ...
function inner(dest, bc, j)
    for i in axes(bc)[1]
        dest[i, j] = bc[i, j]
    end
end
```

Another interesting/crazy thought would be to manually unroll a number (maybe 16?) of loops — does constant propagation save you if you do

```julia
m, n = size(bc)
n == 0 && return dest
for i=1:m; dest[i, 1] = bc[i, 1]; end
n == 1 && return dest
for i=1:m; dest[i, 2] = bc[i, 2]; end
n == 2 && return dest
# and so on...
```

Of course, that's not OffsetArray safe, but do you support offset data frames?

Reply:
Thank you. Just to confirm I understand the idea correctly - for each column you replace an `AbstractDataFrame` with a single column extracted from this data frame and let broadcast mechanics do the rest - right?

Reply:
I don't really understand all the mechanics, but does `bc[i, j]` in the inner loop imply indexing a data frame under the hood, or just the column vector? Is there any way we could add an optimized path which calls `copyto!` (which is much faster for `CategoricalArray` and `PooledArray`) when the source is a column vector in a data frame?

Reply:
I have just pushed out a working implementation of what @mbauman proposal. It is still WIP as I have not tested it end-to-end. Hopefully this will clarify.

However, I have one general note about allowing data frames on right hand side of broadcasting. It is going to be very slow for data frames with a large number of columns. The reason is that we have to unalias all pairs of columns in source and destination. So for `10_000` columns you have to do `100_000_000` checks, which is an overkill. Maybe it would be better to just copy a data frame if it is in source and has a number of columns over some limit?

Reply:
@nalimilan I managed to implement some improvements and now the broadcasting speed is not that bad even for large data frames. But for very large data frames it is better to convert a `DataFrame` into e.g. a `Matrix` as it will be much faster because we are sure we do not have to do do alias analysis.